core/src/main/java/software/amazon/awssdk/core/sync/ResponseBytes.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.sync;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * An in-memory representation of the service's response from a streaming operation. This usually obtained by calling the "bytes"
+ * form of a streaming operation, like S3's {@code getObjectBytes}. Can also be retrieved by passing
+ * {@link StreamingResponseHandler#toBytes()} to a streaming output operation.
+ */
+@SdkPublicApi
+public final class ResponseBytes<ResponseT> {
+    private final ResponseT response;
+    private final byte[] responseBytes;
+
+    public ResponseBytes(ResponseT response, InputStream responseStream) throws IOException {
+        this.response = response;
+        this.responseBytes = IoUtils.toByteArray(responseStream);
+    }
+
+    /**
+     * @return the unmarshalled response object from the service.
+     */
+    public ResponseT response() {
+        return response;
+    }
+
```

Comment:
Would be useful to have an asBytes or asByteArray to return a copy of the byte array.

Reply:
Customers currently have the same problem when they use our models, so I didn't think it would be an issue here. Is it not a problem there as well?

Reply:
Not sure I follow. This is a container for an in memory representation and a bunch of transformation methods to transform it to usable types. So both byte array and byte buffer transformation methods would be valid on this class. I'm not against exploring a different representation for models, I don't find ByteBuffer particularly easy to use and it's much easier to go from byte array to byte buffer than the other way around. This pattern (minus the response container part) might be useful in models too but that's a separate discussion to be had.

Reply:
You seemed to follow just fine! 😄

I'll add one here that makes it clear it will copy the byte array, and we can discuss doing something similar for the models separately.